drivers/timer/apic_timer.c

```diff
@@ -0,0 +1,378 @@
+ */
+
+#define CYCLES_PER_TICK \
+	(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+BUILD_ASSERT_MSG(CYCLES_PER_TICK >= 2, "APIC timer: bad CYCLES_PER_TICK");
+
+/*
+ * MAX_TICKS is max number of ticks we can load into the timer in one shot.
+ */
+
+#define MAX_TICKS (0xFFFFFFFFU / CYCLES_PER_TICK)
+
+/*
+ * The spinlock protects all access to the local APIC timer
+ * registers, as well as 'total_cycles' and 'last_annoucement'.
+ */
+
+static struct k_spinlock lock;
+static u64_t total_cycles;	/* cycle accumulator */
+static u64_t last_announcement;	/* last time we called z_clock_announce() */
+
+/*
+ * The free-running timer referenced by z_timer_cycle_get_32() is, in an
+ * abstract sense, not required to be related to the timer driver proper,
+ * excepting that it must appear to have the same frequency. So, when
+ * appropriate, use the TSC frequency-scaled to provide this timer; when
+ * not available, use the decidedly inferior accumulator approximation.
+ */
+
+#ifdef CONFIG_APIC_TIMER_TSC
+
+u32_t z_timer_cycle_get_32(void)
+{
+	u64_t tsc = z_tsc_read();
+	u32_t cycles;
+
+	cycles = (tsc * CONFIG_APIC_TIMER_TSC_M) / CONFIG_APIC_TIMER_TSC_N;
+	return cycles;
+}
+
+#else
+
+/*
+ * As explained in the theory of operation, 'total_cycles' plus the cycles
+ * counted off is cycles-since-power on, subject to drift. We only return
+ * the lower 32 bits; we're obviously prepared if the API extends this to 64.
+ */
+
+u32_t z_timer_cycle_get_32(void)
+{
+	u32_t ret;
+	u32_t ccr;
+	u32_t icr;
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	ccr = x86_read_loapic(LOAPIC_TIMER_CCR);
+	icr = x86_read_loapic(LOAPIC_TIMER_ICR);
+	ret = total_cycles + (icr - ccr);
+	k_spin_unlock(&lock, key);
+
+	return ret;
+}
+
+#endif
+
+/*
+ * Regardless of mode of operation, the initialization is the same: set
+ * the divisor to 1, one-shot mode, install the ISR, program first tick.
+ * Be aware of the note on step 3: the interrupt code ALSO manipulates
+ * the LOAPIC_TIMER register, so be sure to account for that in any changes.
+ */
+
+int z_clock_driver_init(struct device *device)
+{
+	u32_t val;
+
+	/* 1. set DCR - clock divider */
+
+	val = x86_read_loapic(LOAPIC_TIMER_CONFIG);
+	val &= ~DCR_DIVIDER_MASK;
+	val |= DCR_DIVIDER;
+	x86_write_loapic(LOAPIC_TIMER_CONFIG, val);
+
+	/* 2. set LVT - timer mode */
+
+	val = x86_read_loapic(LOAPIC_TIMER);
+	val &= ~LVT_MODE_MASK;
+	val |= LVT_MODE;
+	x86_write_loapic(LOAPIC_TIMER, val);
+
+	/* 3. wire up interrupt (this will configure remaining LVT bits) */
+
+	IRQ_CONNECT(CONFIG_APIC_TIMER_IRQ,
+		CONFIG_APIC_TIMER_IRQ_PRIORITY,
+		isr, 0, 0);
+	irq_enable(CONFIG_APIC_TIMER_IRQ);
+
+	/* 4. schedule the first interrupt */
+
+	x86_write_loapic(LOAPIC_TIMER_ICR, CYCLES_PER_TICK);
+
+	return 0;
+}
+
+#ifdef CONFIG_TICKLESS_KERNEL
+
+/*
+ * Based on my understanding of the contract described in
+ * include/drivers/system_timer.h, this function acts as follows:
+ *
```

Comment:
Better to update the contract in include/drivers/system_timer.h then than repeat it here, no? :)

Reply:
But that's _your_ documentation. :)